### PR TITLE
SIP: URLs in spin.toml sources

### PR DESCRIPTION
**Pull request by technosophos**
SIP for supporting `source` that uses HTTP/HTTPS or `file` schema.

Example:

```toml
[[component]]
source = "https://github.com/fermyon/bartholomew/releases/download/v0.3.0/bartholomew.wasm#sha256:723ade692c4f8b047b6e90b7c9625a57c9819e0f1bd29bd75eae5b79dd436c4b"
```

Signed-off-by: Matt Butcher <matt.butcher@fermyon.com>

### CODE REVIEW

**Comment by radu-matei** [comment]
Today you _can_ reference a remote parcel from a remote bindle:

```toml
[component.source]
reference = "fermyon/bartholomew/0.3.0"
parcel = "bartholomew.wasm"
```

But this assumes an always running Bindle server with anonymous pulls enabled — which we can definitely do for our projects, but this does not immediately address the needs of people who don't want to run their own.


We have [explored in #393](https://github.com/fermyon/spin/issues/393#issuecomment-1109093546) using a URL pointing to a module and its digest as the component source:

```
[component.source]
url = "https://raw.githubusercontent.com/..."
integrity = "sha256:abc123..."
```

Similar to the Bindle component source, where we have `reference` and `parcel`, I think having separate `url` and `hash`/`integrity` fields is a bit more clear.


TL; DR — this looks good to me, and I'd prefer separate fields for the new component source option.

**Comment by lann** [comment]
ref #393

**Comment by vdice** [comment]
@technosophos now that we have support for fetching wasm modules via the web (courtesy https://github.com/fermyon/spin/pull/890), should we close this PR?  There is one piece of functionality mentioned in this SIP that isn't yet in Spin, caching functionality, but we do have the following open issue around this: https://github.com/fermyon/spin/issues/455.  As caching is broader than the particular mechanism for fetching modules (eg bindle, web URL, OCI reference), it makes sense to me to track separately anyhow.  WDYT?

**Comment by itowlson** [comment]
@vdice This SIP also asks that we support SHA512, redirects, `file:` URLs, and omitting the digest, though of course we can raise separate issues for those (or may feel that they aren't immediate enough to need capturing now).

**Comment by technosophos** [comment]
When you read a PR and have NO RECOLLECTION of writing it....

Yeah, capturing the few related items as issues is probably all that is necessary before closing this. Thanks!

**Comment by vdice** [comment]
Alright, I believe I've tracked all of the follow-ups, linking back to this issue and/or specific sections of the SIP.  These should be bite-sized enough that we wouldn't need to break out their implementation details into their own SIPs (or a revised SIP based on this PR), but I'll defer to @itowlson and @technosophos.  Thanks for adding all of the high-level details on these features here, @technosophos !

**Comment by itowlson** [comment]
Pretty sure we've banned @technosophos from GitHub by this point so I'll use my powers to close it. :trollface:
